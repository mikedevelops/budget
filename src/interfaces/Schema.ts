export type EntrySchema = ICurrencyEntrySchema | IDateEntrySchema;

export interface ICurrencyEntrySchema {
    type: 'CURRENCY';
    value: 'INT';
    currency: ['Â£'];
    label: 'Currency';
    placeholder: string;
    name: string;
}

export interface IDateEntrySchema {
    type: 'DATE';
    value: 'DATE';
    label: 'Date';
    placeholder: string;
    name: string;
}
