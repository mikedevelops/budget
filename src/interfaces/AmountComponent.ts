export interface AmountComponentPropsInterface {
    currency: string;
    value: number;
}
