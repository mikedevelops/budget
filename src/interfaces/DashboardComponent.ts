export interface DashboardComponentProps {
    targetCurrency: string;
    targetValue: number;
}
