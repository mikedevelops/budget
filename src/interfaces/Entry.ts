export interface ICurrencyEntrySchema {
    type: 'CURRENCY';
    value: 'INT';
    currency: 'STRING';
}
