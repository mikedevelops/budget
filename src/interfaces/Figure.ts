export type Figure = ICurrencyFigure;

export interface ICurrencyFigure {
    type: 'currency';
    value: number;
    meta: {
        currency: 'Â£'
    };
}

export interface IFigureDictionary {
    [index: string]: Figure;
}
