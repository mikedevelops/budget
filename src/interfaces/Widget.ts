export interface ICurrencyWidgetSchema {
    type: 'CURRENCY';
    value: 'INT';
    currency: 'STRING';
}

export interface ICurrencyWidget {
    type: string;
    currency: string;
    value: number;
}
