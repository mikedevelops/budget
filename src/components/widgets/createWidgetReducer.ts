import { Reducer, AnyAction } from 'redux';
import { schema } from '../../schema/schema';
import { ICreateWidgetAction } from './createWidgetActions';
import { WidgetSchema } from '../../interfaces/Schema';

const defaultState: any = {
    schema,
    widgets: []
};

// MEGA TODO: Create functions for creating widet objects form widget schema

const createWidgetReducer: Reducer<any> = (state: any = defaultState, action: any) => {
    switch (action.type) {
        case 'CREATE_WIDGET':
            const { formData } = action as ICreateWidgetAction;

            // TODO: Should we validate these props are available

            const blueprint = state.schema.find((s: WidgetSchema) =>
                s.type.toLowerCase() === formData.schema.toLowerCase());

            // TODO: Should this switch on formDate.schema and build widget objects

            if (blueprint === undefined) {
                throw new Error(`Could not find schema for "${formData.schema}"`);
            }

            // TODO: Widget objects should be built functionaly

            const widget = Object.assign({}, blueprint, {
                name: formData[`${formData.schema}_name`]
            });

            // TODO: factor this out of the reducer

            Object.keys(widget).reduce((w: any, key: string) => {
                for (const dataKey in formData) {
                    if (formData.hasOwnProperty(dataKey)) {
                        const widgetKey = dataKey.split('_')[1];

                        if (w[widgetKey] !== undefined) {
                            w[widgetKey] = formData[dataKey];
                        }
                    } else {
                        console.warn(`Form data for "${widget}" contains data "${key}" that does not exist in schema`);
                    }
                }

                return w;
            }, widget);

            return Object.assign({}, state, {
                widgets: [...state.widgets, widget]
            });
        default:
            return state;
    }
};

export default createWidgetReducer;
