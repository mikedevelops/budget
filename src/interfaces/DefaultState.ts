import { IFigureDictionary } from './Figure';

export interface IDefaultState {
    entries: IFigureDictionary[];
}
