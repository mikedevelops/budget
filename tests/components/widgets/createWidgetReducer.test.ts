import {
    ICreateWidgetAction,
    createWidgetAction
} from '../../../src/components/widgets/createWidgetActions';
import createWidgetReducer from '../../../src/components/widgets/createWidgetReducer';
import { AnyAction } from 'redux';
import { ICurrencyWidgetSchema } from '../../../src/interfaces/Schema';
import { ICurrencyWidget } from '../../../src/interfaces/Widget';

describe('createWidgetReducer', () => {
    let state: any;

    beforeEach(() => {
        state = {
            schema: [
                {
                    type: 'CURRENCY',
                    value: 'INT',
                    currency: ['Â£'],
                    label: 'Currency',
                    placeholder: '0.00',
                    name: ''
                }
            ],
            widgets: []
        };
    });

    describe('UPDATE_SCHEMA', () => {
        test('should throw an error if the schema does not have a blueprint', () => {
            const action: AnyAction = {
                type: 'A_SCHEMA_THAT_DOES_NOT_EXIST'
            };

            expect(createWidgetReducer(state, action)).toThrowErrorMatchingSnapshot();
        });

        test('should insert a new widget to the state', () => {
            const action: ICreateWidgetAction = createWidgetAction({
                schema: 'CURRENCY',
                foo: 'bar'
            });

            const widget: ICurrencyWidget = {
                type: 'CURRENCY',
                currency: '$',
                value: 666
            };

            expect(createWidgetReducer(state, action)).toEqual({
                schema: state.schema,
                widgets: [widget]
            });
        });
    });
});
