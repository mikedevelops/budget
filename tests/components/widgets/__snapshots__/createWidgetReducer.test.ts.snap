// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createWidgetReducer UPDATE_SCHEMA should throw an error if the schema does not have a blueprint 1`] = `"received is not a function"`;
